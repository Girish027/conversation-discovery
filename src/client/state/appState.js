import {
  fromJS, List, Map
} from 'immutable';
import { createSelector } from 'reselect';
import axios from 'axios';
import _ from 'lodash';
import { push } from 'connected-react-router';
import { getURLParams } from 'utils/URIUtils';
import { routeNames, getRoute } from 'utils/RouteHelper';
import { getPathnameSelector } from 'state/routerState';
import appLogger from 'utils/appLogger';
import { LOCATION_CHANGE, UPDATE_CAA } from 'state/types';
import { getAllProjects } from 'state/projectsState';
import { setDefaultsForAPI } from 'utils/api';
import { isUserAuthorized, getSortedClientsList } from 'utils/caaHelper';
import constants from '../constants';
import { clearStorage, loadStorage, saveStorage } from '../utils/storageManager';
import { logAmplitudeEvent } from '../utils/amplitudeUtils';
import { roles, updateRoleConfig } from '../utils/roleHelper';

export const types = {
  SET_USER_CONFIG: 'app/SET_USER_CONFIG',
  SET_CLIENTS_CONFIG: 'app/SET_CLIENTS_CONFIG',
  MODAL_IS_OPEN: 'app/MODAL_IS_OPEN',
  MODAL_MESSAGE: 'app/MODAL_MESSAGE',
  INIT_ANALYTICS: 'app/INIT_ANALYTICS',
  SET_UNIFIED_PORTAL_CLIENT_PICKER: 'app/SET_UNIFIED_PORTAL_CLIENT_PICKER',
  SHOW_WORD_CLOUD: 'app/SHOW_WORD_CLOUD',
};

export const stateKey = {
  authEnabled: 'authEnabled',
  clientId: 'clientId',
  componentClientId: 'componentClientId',
  appId: 'appId',
  accountId: 'accountId',
  userName: 'userName',
  userType: 'userType',
  itsBaseUrl: 'itsBaseUrl',
  analyticsKey: 'analyticsKey',
  clientConfigs: 'clientConfigs',
  modalIsOpen: 'modalIsOpen',
  modalMessage: 'modalMessage',
  modalState: 'modalState',
  unifiedPortalClientPicker: 'unifiedPortalClientPicker',
  docPortalUrl: 'docPortalUrl',
  unifiedPortalUrl: 'unifiedPortalUrl',
  contactSupportUrl: 'contactSupportUrl',
  oktaUserAccountUrl: 'oktaUserAccountUrl',
  isAnswers: 'isAnswers',
  isConversations: 'isConversations',
  userRole: 'userRole',
  roleConfig: 'roleConfig',
  showWordCloud: 'showWordCloud',
  wordCloudcontent: 'wordCloudcontent',
};

export const initialAppState = fromJS({
  authEnabled: false,
  clientId: '',
  componentClientId: '',
  appId: '',
  accountId: '',
  userName: '',
  userType: '',
  itsBaseUrl: '',
  analyticsKey: '',
  clientConfigs: List(),
  modalIsOpen: false,
  modalMessage: '',
  modalState: {},
  unifiedPortalClientPicker: false,
  docPortalUrl: '',
  unifiedPortalUrl: '',
  contactSupportUrl: '',
  oktaUserAccountUrl: '',
  isConversations: false,
  isAnswers: false,
  userRole: '',
  roleConfig: {},
  showWordCloud: false,
  wordCloudcontent: '',
});

const devlog = appLogger({ tag: 'appState' });

export const updateCAA = (caa) => ({
  type: UPDATE_CAA,
  caa,
});

export const setUnifiedPortalClientPicker = (unifiedPortalClientPicker) => ({
  type: types.SET_UNIFIED_PORTAL_CLIENT_PICKER,
  unifiedPortalClientPicker
});

export const setUserConfig = (config) => ({
  type: types.SET_USER_CONFIG,
  config,
});

export const setClientsConfig = (clientConfigs = []) => ({
  type: types.SET_CLIENTS_CONFIG,
  clientConfigs,
});

export const setShowWordCloud = (showWordCloud = false, wordCloudcontent = '') => ({
  type: types.SHOW_WORD_CLOUD,
  showWordCloud,
  wordCloudcontent,
});

export const setModalIsOpen = (modalIsOpen, modalState = {}) => ({
  type: types.MODAL_IS_OPEN,
  modalIsOpen,
  modalState,
});

export const setModalMessageState = (modalMessage) => ({
  type: types.MODAL_MESSAGE,
  modalMessage
});

export const setModalMessage = (modalMessage) => (dispatch) => {
  dispatch(setModalMessageState(modalMessage));
};

export const getConversationsConfig = () => (dispatch, getState) => {
  axios.get(constants.serverApiUrls.conversationsConfigUrl, { withCredentials: true })
    .then((configRes) => {
      const { data: configBody = {} } = configRes;
      const conversationsClientConfigs = configBody.clientConfigs;
      let clientConfigs = getClientConfigsSelector(getState());
      clientConfigs = clientConfigs.map((element) => {
        let configObj = element;
        const { clientId } = element;
        const conversationsClient = conversationsClientConfigs.find((config) => config.clientId === clientId);
        if (conversationsClient !== undefined) {
          configObj = { ...configObj, isConversations: true };
        }
        return configObj;
      });
      dispatch(setClientsConfig(clientConfigs));
    })
    .catch((configErr) => {
      devlog.error(configErr);
    });
};

export const getAnswersConfig = () => (dispatch, getState) => {
  axios.get(constants.serverApiUrls.answersConfigUrl, { withCredentials: true })
    .then((configRes) => {
      const { data: configBody = {} } = configRes;
      const answersClientConfigs = configBody.clientConfigs;
      let clientConfigs = getClientConfigsSelector(getState());
      clientConfigs = clientConfigs.map((element) => {
        let configObj = element;
        const { clientId } = element;
        const answersClient = answersClientConfigs.find((config) => config.clientId === clientId);
        if (answersClient !== undefined) {
          configObj = { ...configObj, isAnswers: true };
        }
        return configObj;
      });
      dispatch(setClientsConfig(clientConfigs));
    })
    .catch((configErr) => {
      devlog.error(configErr);
    });
};

export const getUserConfig = () => (dispatch, getState) => {
  axios.get(constants.serverApiUrls.userConfigUrl, { withCredentials: true })
    .then((configRes) => {
      const { data: configBody = {} } = configRes;
      setDefaultsForAPI();

      const sortClientConfigs = getSortedClientsList([...configBody.clientConfigs]);
      dispatch(setClientsConfig(sortClientConfigs));
      dispatch(getAnswersConfig());
      dispatch(getConversationsConfig());
      const userSelectedCAA = getCAASelector(getState());

      const { authorized, clientPicker, caa } = isUserAuthorized(userSelectedCAA, sortClientConfigs);
      if (clientPicker) {
        dispatch(setUnifiedPortalClientPicker(true));
      } else if (!authorized && (userSelectedCAA.clientId !== 'undefined' || userSelectedCAA.appId !== 'undefined')) {
        dispatch(setModalIsOpen(true, {
          modalName: constants.modals.unauthorized,
          header: 'Access Denied',
          message: `You are not authorized to view the client: ${userSelectedCAA.clientId}, app: ${userSelectedCAA.appId} and account: ${userSelectedCAA.accountId}`,
          onClickOk: () => dispatch(closeUnauthorizedModal()),
          onClickClose: () => dispatch(closeUnauthorizedModal()),
          onClickCancel: () => dispatch(closeUnauthorizedModal()),
        }));
      } else {
        dispatch(changeClient(caa));
      }
      dispatch(setUserConfig(configBody));
    })
    .catch((configErr) => {
      devlog.error(configErr);
    });
};

const closeUnauthorizedModal = () => (dispatch) => {
  dispatch(setUnifiedPortalClientPicker(true)); dispatch(setModalIsOpen(false));
};

// When user changes client
export const changeClient = (caa) => (dispatch, getState) => {
  const state = getState();
  const existingCAA = getCAASelector(state);
  if (!_.isEqual(existingCAA, caa)) {
    dispatch(updateCAA(caa));
    clearStorage(constants.activeProjectId);
    clearStorage(constants.isConversations);
    clearStorage(constants.isAnswers);
    clearStorage(constants.userRole);
  }
  const { clientId, appId } = caa;
  const obj = {
    clientId, appId, toolId: constants.toolName, env: constants.environment
  };
  logAmplitudeEvent('IDTClients', obj);
  dispatch(loadRoute(caa));
  dispatch(getAllProjects(caa));
};

const loadRoute = (caa) => (dispatch, getState) => {
  // refresh the page only if it is HOMEPAGE/baseRoute.
  const state = getState();
  const currentPath = getPathnameSelector(state);
  if (currentPath === `${routeNames.HOME_PAGE}/` || currentPath === `${routeNames.BASE_ROUTE}/` || currentPath === `${routeNames.DISCOVER_INTENTS}`) {
    const projectId = loadStorage(constants.activeProjectId, '');
    const newRoute = getRoute(routeNames.DISCOVER_INTENTS, { ...caa, projectId });
    dispatch(push(newRoute));
  }
};

// REDUCERS
export const appReducer = (state = initialAppState, action) => {
  switch (action.type) {
    case types.SET_CLIENTS_CONFIG:
      return state.set(stateKey.clientConfigs, List(action.clientConfigs));
    case types.SET_USER_CONFIG:
      return state.merge(Map({
        userName: action.config.username,
        userType: action.config.userType,
        authEnabled: action.config.authEnabled,
        itsBaseUrl: action.config.itsBaseUrl,
        analyticsKey: action.config.analyticsKey,
        docPortalUrl: action.config.docPortalUrl,
        unifiedPortalUrl: action.config.unifiedPortalUrl,
        contactSupportUrl: action.config.contactSupportUrl,
        oktaUserAccountUrl: action.config.oktaUserAccountUrl,
      }));
    case types.SET_UNIFIED_PORTAL_CLIENT_PICKER:
      return state.set(stateKey.unifiedPortalClientPicker, action.unifiedPortalClientPicker);
    case types.MODAL_IS_OPEN:
      return state.merge(Map({
        modalIsOpen: action.modalIsOpen,
        modalState: action.modalState,
      }));
    case types.MODAL_MESSAGE:
      return state.set(stateKey.modalMessage, action.modalMessage);
    case types.SHOW_WORD_CLOUD:
      return state.merge(Map({
        showWordCloud: action.showWordCloud,
        wordCloudcontent: action.wordCloudcontent,
      }));
    // The below reducers are handling part of the global state change
    case UPDATE_CAA:
      return state.merge(Map(action.caa));
    case LOCATION_CHANGE: {
      const { search } = action.payload.location;
      if (search) {
        const {
          clientid, appid, accountid, componentClientId
        } = getURLParams(search);

        let isAnswers = false;
        let isConversations = false;

        const client = state.get(stateKey.clientId);
        const clientConfigs = state.get(stateKey.clientConfigs);
        const selectedClientConfig = clientConfigs.find((config) => config.clientId === clientid);
        let userRole = roles.unauthorized;
        if (selectedClientConfig) {
          userRole = selectedClientConfig.role;
        }

        if (selectedClientConfig && selectedClientConfig.isAnswers) {
          isAnswers = true;
        }
        if (selectedClientConfig && selectedClientConfig.isConversations) {
          isConversations = true;
        }

        if (client === '') {
          isConversations = loadStorage(constants.isConversations, false);
          isAnswers = loadStorage(constants.isAnswers, false);
          userRole = loadStorage(constants.userRole, roles.unauthorized);
        } else {
          saveStorage(constants.isConversations, isConversations);
          saveStorage(constants.isAnswers, isAnswers);
          saveStorage(constants.userRole, userRole);
        }
        const roleConfig = updateRoleConfig(userRole);
        return state.merge(Map({
          clientId: clientid,
          appId: appid,
          accountId: accountid,
          componentClientId,
          isAnswers,
          isConversations,
          userRole,
          roleConfig,
        }));
      }
      return state;
    }
    default:
      return state;
  }
};

// SELECTORS
export const getAppState = (state) => state.get('app');

export const getCAASelector = createSelector(
  getAppState,
  (appState) => ({
    clientId: appState.get(stateKey.clientId),
    appId: appState.get(stateKey.appId),
    accountId: appState.get(stateKey.accountId),
    componentClientId: appState.get(stateKey.componentClientId),
  })
);

export const getClientIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.clientId
);

export const getComponentClientIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.componentClientId
);

export const getAppIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.appId
);

export const getIsAnswersSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.isAnswers)
);

export const getIsConversationsSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.isConversations)
);

export const getAccountIdSelector = createSelector(
  getCAASelector,
  (caa) => caa.accountId
);

export const getClientConfigsSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.clientConfigs)
);

export const getITSBaseUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.itsBaseUrl)
);

export const getShowWordCloudSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.showWordCloud)
);

export const getWordCloudContentSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.wordCloudcontent)
);

export const getModalOpenSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.modalIsOpen)
);

export const getModalStateSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.modalState)
);

export const getModalMessageSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.modalMessage)
);

export const getUnifiedPortalClientPicker = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.unifiedPortalClientPicker)
);

export const getDocPortalUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.docPortalUrl)
);

export const getUnifiedPortalUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.unifiedPortalUrl)
);

export const getUserTypeSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.userType)
);

export const getContactSupportUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.contactSupportUrl)
);

export const getUserRoleSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.userRole)
);

export const getRoleConfigSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.roleConfig)
);

export const getOktaUserAccountUrlSelector = createSelector(
  getAppState,
  (appState) => appState.get(stateKey.oktaUserAccountUrl)
);
