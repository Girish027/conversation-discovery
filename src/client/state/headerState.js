import {
  fromJS
} from 'immutable';
import Constants from '../constants';

// Implement:
// 1. Curation has clickable links in notification. Current TFSUI implementation is limited to just strings.

export const types = {
  HIDE_NOTIFICATION: 'CLEAR_ALL_NOTIFICATION',
  SHOW_NOTIFICATION: 'header/SHOW_NOTIFICATION',
};

export const initialHeaderState = fromJS({
  message: '',
  notificationType: ''
});

// ACTIONS
export const showNotificationWithTimer = (message, notificationType = Constants.notification.info, notificationInterval = Constants.notificationInterval) => (dispatch) => {
  dispatch(showNotification(message, notificationType));
  setTimeout(() => {
    dispatch(hideNotification());
  }, notificationInterval);
};

export const showNotification = (message, notificationType) => ({
  type: types.SHOW_NOTIFICATION,
  message,
  notificationType,
});

export const hideNotification = () => ({
  type: types.HIDE_NOTIFICATION,
});

// REDUCERS
export const headerReducer = (state = initialHeaderState, action) => {
  switch (action.type) {
    case types.HIDE_NOTIFICATION: return initialHeaderState;
    case types.SHOW_NOTIFICATION: {
      return Object.assign({}, state, {
        message: action.message,
        notificationType: action.notificationType,
      });
    }
    default:
      return state;
  }
};

// SELECTORS
export const getHeaderState = (state) => state.get('header');

export const getNotifications = (state) => getHeaderState(state);

// SELECTORS TO BE USED IN COMPONENTS
export const headerSelectors = {
  getNotifications
};


// ACTIONS TO BE DISPATCHED FROM COMPONENTS
export const headerActions = {
  showNotificationWithTimer
};
