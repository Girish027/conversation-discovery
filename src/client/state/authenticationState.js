import {
  fromJS, Map
} from 'immutable';
import { createSelector } from 'reselect';
import axios from 'axios'; // this module is Promise based HTTP client for the browser and node.js
import appLogger from 'utils/appLogger';
import { getUserConfig } from 'state/appState';
import constants from '../constants';
import { initializeAnalytics, logAmplitudeEvent } from '../utils/amplitudeUtils';

export const types = {
  AUTHENTICATION_COMPLETE: 'auth/AUTHENTICATION_COMPLETE',
  LOGGED_OUT: 'auth/LOGGED_OUT'
};

export const stateKey = {
  name: 'name',
  oktaUrl: 'oktaUrl',
  email: 'email'
};

const devlog = appLogger({ tag: 'authenticationState' });

export const initialAuthState = fromJS({
  name: 'User Name',
  oktaUrl: ''
});

export const actionCompletedAuth = (userInfo) => ({
  type: types.AUTHENTICATION_COMPLETE,
  userInfo
});

export const actionLoggedoutAuth = () => ({
  type: types.LOGGED_OUT
});

export const getAuthUserinfo = () => (dispatch) => {
  axios.get(constants.serverApiUrls.authentications)
    .then((res) => {
      const authInfoArray = res.data;
      if (Array.isArray(authInfoArray) && authInfoArray.length) {
        const authInfo = authInfoArray[0];
        const userInfo = {
          name: authInfo.name,
          oktaUrl: authInfo.oktaUrl,
          email: authInfo.email,
        };
        initializeAnalytics(authInfo.analyticsKey, authInfo.name);
        const IDTUserEvent = { username: authInfo.name, toolId: constants.toolName, env: constants.environment };
        logAmplitudeEvent('IDTUser', IDTUserEvent);
        dispatch(actionCompletedAuth(userInfo));
        dispatch(getUserConfig());
      } else {
        dispatch(actionLoggedoutAuth());
      }
    })
    .catch((fetchErr) => {
      devlog.error(fetchErr);
    });
};

// REDUCER
export const authReducer = (state = initialAuthState, action = {}) => {
  switch (action.type) {
    case types.AUTHENTICATION_COMPLETE:
      return state.merge(Map(action.userInfo));
    case types.LOGGED_OUT:
      return initialAuthState;
    default:
      return state;
  }
};

// SELECTORS
export const getAuthState = (state) => state.get('auth');

export const getAuthNameSelector = createSelector(
  getAuthState,
  (authState) => authState.get(stateKey.name),
);

export const getOktaUrlSelector = createSelector(
  getAuthState,
  (authState) => authState.get(stateKey.oktaUrl),
);

export const getEmailSelector = createSelector(
  getAuthState,
  (authState) => authState.get(stateKey.email),
);
