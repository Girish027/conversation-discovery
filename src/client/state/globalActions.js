import { downloadFile } from 'utils/FileUtils';
import { getCAASelector } from 'state/appState';
import { getActiveProjectId } from 'state/projectsState';
import Constants from '../constants';
import { logAmplitudeEvent } from '../utils/amplitudeUtils';
import {
  DOWNLOAD_FILE
} from './types';

export const downloadResults = (constructURL, runId, clusterId = '') => (dispatch, getState) => {
  const state = getState();
  const projectId = getActiveProjectId(state);
  const caa = getCAASelector(state);
  const locationUrl = constructURL({
    ...caa, projectId, runId, clusterId
  });
  dispatch({ type: DOWNLOAD_FILE, locationUrl });
  const downloadRun = { toolId: Constants.toolName, runId, env: Constants.environment };
  logAmplitudeEvent('IDT_DownloadRunResults', downloadRun);
  downloadFile(locationUrl);
};
