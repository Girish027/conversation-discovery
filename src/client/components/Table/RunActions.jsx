import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  Xmark,
  Pencil,
  Download
} from '@tfs/ui-components';
import IconButton from 'components/IconButton';
import {
  setModalIsOpen,
} from 'state/appState';
import Constants from '../../constants';
import { runActions } from 'state/runsState';
import * as globalActions from 'state/globalActions';

class RunActions extends Component {
  constructor() {
    super();

    this.iconProps = {
      padding: '5px',
      height: 12,
      width: 12,
    };
  }

  onClickEdit = (event) => {
    event.stopPropagation();
    const { dispatch, runId, ...runInfo } = this.props;
    dispatch(setModalIsOpen(true, {
      modalName: Constants.modals.editRun,
      runId,
      formData: runInfo,
    }));
  }

  onClickConfirmDelete = () => {
    const { dispatch, runId } = this.props;
    dispatch(runActions.deleteRun(runId));
  }

  onClickDelete = (event) => {
    event.stopPropagation();
    const { dispatch, runName } = this.props;
    dispatch(setModalIsOpen(true, {
      modalName: Constants.modals.confirm,
      deleteMode: true,
      header: `Delete "${runName}"?`,
      message: 'These analysis results will be gone forever. Are you sure you want to delete this item?',
      onClickOk: this.onClickConfirmDelete,
    }));
  }

  onClickDownload = (event) => {
    event.stopPropagation();
    const { dispatch, runId } = this.props;
    dispatch(globalActions.downloadResults(Constants.serverApiUrls.downloadRun, runId));
  }

  renderDeleteIcon = () => {
    const { runStatus } = this.props;
    let disabled = false;
    switch (runStatus) {
      case Constants.status.QUEUED:
      case Constants.status.IN_PROGRESS:
        disabled = true;
        break;
      default:
        break;
    }
    return (
      <IconButton
        onClick={this.onClickDelete}
        icon={Xmark}
        data-qa={'delete-run'}
        title='Delete'
        disabled={disabled}
        {...this.iconProps}
      />
    );
  }

  renderEditIcon = () => (
    <IconButton
      onClick={this.onClickEdit}
      icon={Pencil}
      defaultColor='none'
      strokeColor='#004c97'
      data-qa={'edit-run'}
      title='Edit'
      {...this.iconProps}
    />
  )

  renderDownloadIcon = () => {
    const { runStatus, projectType } = this.props;
    let disabled = true;
    if (runStatus === Constants.status.READY && projectType !== Constants.projects.SYSTEM) {
      disabled = false;
    }
    return (
      <IconButton
        onClick={this.onClickDownload}
        icon={Download}
        data-qa={'delete-run'}
        title='Download Run Results'
        disabled={disabled}
        {...this.iconProps}
      />
    );
  }

  render() {
    const { roleConfig } = this.props;
    return (
      <div className='float-left'>
        { roleConfig.deleteAnalysis ? this.renderDeleteIcon() : null }
        {this.renderDownloadIcon()}
        { roleConfig.updateAnalysis ? this.renderEditIcon() : null }
      </div>
    );
  }
}

RunActions.propTypes = {
  runName: PropTypes.string.isRequired,
  runDescription: PropTypes.string,
  runId: PropTypes.string.isRequired,
  runStatus: PropTypes.string.isRequired,
  dispatch: PropTypes.func,
  projectType: PropTypes.string,
  roleConfig: PropTypes.object,
};

RunActions.defaultProps = {
  runDescription: '',
  roleConfig: {},
};

export default RunActions;
