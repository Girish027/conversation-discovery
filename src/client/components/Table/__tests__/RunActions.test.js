import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import RunActions from 'components/Table/RunActions';
import { runActions } from 'state/runsState';
import * as appState from 'state/appState';
import * as globalActions from 'state/globalActions';
import Constants from 'components/../constants';

jest.mock('state/globalActions');
jest.mock('state/appState');

describe('<RunActions />', function () {
  let wrapper;

  const props = {
    runName: 'name',
    runDescription: 'desc',
    runId: 'run-123',
    runStatus: 'QUEUED',
  };

  afterAll(() => {
    jest.clearAllMocks();
  });

  const getShallowWrapper = (propsObj) => shallow(<RunActions
    {...propsObj}
  />);

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = getShallowWrapper(props);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly for the given props - status QUEUED', () => {
      wrapper = getShallowWrapper(props);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('renders correctly for the props - READY', () => {
      wrapper = getShallowWrapper({ ...props, runStatus: 'READY' });
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('renders correctly for the props - FAILED', () => {
      wrapper = getShallowWrapper({ ...props, runStatus: 'FAILED' });
      expect(toJSON(wrapper)).toMatchSnapshot();
    });

    test('renders correctly for the props - IN PROGRESS', () => {
      wrapper = getShallowWrapper({ ...props, runStatus: 'IN PROGRESS' });
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Functionality:', () => {
    beforeEach(function () {
      props.dispatch = jest.fn();
      runActions.deleteRun = jest.fn(() => 'deleteRun');
      globalActions.downloadResults = jest.fn();
      appState.setModalIsOpen = jest.fn();
    });

    afterEach(() => {
      jest.clearAllMocks();
    });

    test('onClickEdit', () => {
      const event = {
        stopPropagation: () => {}
      };
      appState.setModalIsOpen.mockReturnValue('setModalIsOpen');
      wrapper = getShallowWrapper(props);
      wrapper.instance().onClickEdit(event);
      expect(appState.setModalIsOpen).toHaveBeenCalledWith(true, expect.any(Object));
      expect(appState.setModalIsOpen.mock.calls[0][1]).toMatchSnapshot();
      expect(props.dispatch).toHaveBeenCalledWith('setModalIsOpen');
    });

    test('onClickConfirmDelete', () => {
      wrapper = getShallowWrapper(props);
      wrapper.instance().onClickConfirmDelete();
      expect(runActions.deleteRun).toHaveBeenCalledWith(props.runId);
      expect(props.dispatch).toHaveBeenCalledWith('deleteRun');
    });

    test('onClickDelete', () => {
      const event = {
        stopPropagation: () => {}
      };
      appState.setModalIsOpen.mockReturnValue('setModalIsOpen');
      wrapper = getShallowWrapper(props);
      wrapper.instance().onClickDelete(event);
      expect(appState.setModalIsOpen).toHaveBeenCalledWith(true, expect.any(Object));
      expect(appState.setModalIsOpen.mock.calls[0][1]).toMatchSnapshot();
      expect(props.dispatch).toHaveBeenCalledWith('setModalIsOpen');
    });

    test('onClickDownload', () => {
      const event = {
        stopPropagation: () => {}
      };
      globalActions.downloadResults.mockReturnValue('downloadResults');
      wrapper = getShallowWrapper(props);
      wrapper.instance().onClickDownload(event);
      expect(globalActions.downloadResults).toHaveBeenCalledWith(Constants.serverApiUrls.downloadRun, props.runId);
      expect(props.dispatch).toHaveBeenCalledWith('downloadResults');
    });
  });
});
