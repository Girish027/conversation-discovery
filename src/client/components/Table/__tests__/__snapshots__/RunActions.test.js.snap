// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RunActions /> Functionality: onClickDelete 1`] = `
Object {
  "deleteMode": true,
  "header": "Delete \\"name\\"?",
  "message": "These analysis results will be gone forever. Are you sure you want to delete this item?",
  "modalName": "ConfirmationDialog",
  "onClickOk": [Function],
}
`;

exports[`<RunActions /> Functionality: onClickEdit 1`] = `
Object {
  "formData": Object {
    "roleConfig": Object {},
    "runDescription": "desc",
    "runName": "name",
    "runStatus": "QUEUED",
  },
  "modalName": "EditRunModal",
  "runId": "run-123",
}
`;

exports[`<RunActions /> Snapshots renders correctly for the given props - status QUEUED 1`] = `
<div
  className="float-left"
>
  <IconButton
    data-qa="delete-run"
    defaultColor="#004c97"
    disabled={true}
    disabledColor="#9B9B9B"
    focusedColor="#003467"
    height={12}
    icon={[Function]}
    onClick={[Function]}
    padding="5px"
    strokeColor="none"
    title="Download Run Results"
    width={12}
  />
</div>
`;

exports[`<RunActions /> Snapshots renders correctly for the props - FAILED 1`] = `
<div
  className="float-left"
>
  <IconButton
    data-qa="delete-run"
    defaultColor="#004c97"
    disabled={true}
    disabledColor="#9B9B9B"
    focusedColor="#003467"
    height={12}
    icon={[Function]}
    onClick={[Function]}
    padding="5px"
    strokeColor="none"
    title="Download Run Results"
    width={12}
  />
</div>
`;

exports[`<RunActions /> Snapshots renders correctly for the props - IN PROGRESS 1`] = `
<div
  className="float-left"
>
  <IconButton
    data-qa="delete-run"
    defaultColor="#004c97"
    disabled={true}
    disabledColor="#9B9B9B"
    focusedColor="#003467"
    height={12}
    icon={[Function]}
    onClick={[Function]}
    padding="5px"
    strokeColor="none"
    title="Download Run Results"
    width={12}
  />
</div>
`;

exports[`<RunActions /> Snapshots renders correctly for the props - READY 1`] = `
<div
  className="float-left"
>
  <IconButton
    data-qa="delete-run"
    defaultColor="#004c97"
    disabled={false}
    disabledColor="#9B9B9B"
    focusedColor="#003467"
    height={12}
    icon={[Function]}
    onClick={[Function]}
    padding="5px"
    strokeColor="none"
    title="Download Run Results"
    width={12}
  />
</div>
`;
